### PR TITLE
[StatusIndicator] Add method to get indicator icon screen rect.

### PR DESCRIPTION
**Pull request by bruvzg**
Add `status_indicator_get_rect` method to get indicator rectangle in the screen coordinates. Usefully of positioning popups (existing methods only include mouse click coordinates).

### CODE REVIEW

**Comment by Mickeon** [comment]
I don't quite understand what this status indicator is, could you provide an example?

**Comment by bruvzg** [comment]
> I don't quite understand what this status indicator is

System tray icons (`StatusIndicator` class):

image

File: doc/classes/DisplayServer.xml [hunk 1]
```diff
@@ -1167,12 +1167,21 @@
 				[b]Note:[/b] This method is implemented on Android, iOS, macOS, Windows, and Linux (X11/Wayland).
 			</description>
 		</method>
+		<method name="status_indicator_get_rect" qualifiers="const">
+			<return type="Rect2" />
+			<param index="0" name="id" type="int" />
+			<description>
+				Returns the status indicator rectangle in the screen coordinates. If status indicator is not visible, returns empty rect.
```
Comment (Mickeon) [inline doc/classes/DisplayServer.xml:1174]:
```suggestion
				Returns the rectangle for the given status indicator [param id] in screen coordinates. If the status indicator is not visible, returns an empty [Rect2].
```

File: doc/classes/StatusIndicator.xml [hunk 2]
```diff
@@ -8,6 +8,14 @@
 	</description>
 	<tutorials>
 	</tutorials>
+	<methods>
+		<method name="get_rect" qualifiers="const">
+			<return type="Rect2" />
+			<description>
+				Returns the status indicator rectangle in the screen coordinates. If status indicator is not visible, returns empty rect.
```
Comment (Mickeon) [inline doc/classes/StatusIndicator.xml:15]:
```suggestion
				Returns the status indicator rectangle in screen coordinates. If this status indicator is not visible, returns an empty [Rect2].
```

**Review by Mickeon — COMMENTED** [review]
Aah, roger that.

**Review by Mickeon — APPROVED** [review]
Docs are just fine now, can't speak on implementation

**Review by akien-mga — APPROVED** [review]
Code looks good.

**Comment by Mickeon** [comment]
Can't think of any on top of my head. If not for a popup it could be used for other visual flair (e.g. item inside a window moving towards the taskbar)

**Comment by bruvzg** [comment]
> Is there another usecase, outside of a popup? I wonder if we could perhaps add just an open_popup method of sorts

The only info `org.kde.StatusNotifierItem` provide is position hint (`the x and y parameters are in screen coordinates and is to be considered an hint to the item about where to show the context menu`), and existing `StatusIndicator` callback already include click coordinates (currently it's absolute mouse position, but probably it would make sense to switch it to the same position hint instead for consistency, but I'm not sure what's point it should return exactly). But ideally you should know both tray icon rect and popup rect to position it perfectly, I do not think there's any workaround.

> as I don't see this playing well with Wayland (when/if it will be implemented) at least.

Until full multi-window support is implemented, it's going to be completely useless for Wayland anyway. But I plan to implement `org.kde.StatusNotifierItem` (since it will work with X11 as well).

**Comment by bruvzg** [comment]
> if there's an API for making "native" menus.

Most apps (at least on macOS and Windows) aren't using menus as popup (2/3 of indicators I currently have on macOS are fully custom popups). Not sure if there's any kind of native menu.

> I'm only slightly concerned about the amount of duplicated/deprecated API surface we might have if we ever switch to (or just add) a window-centric API, which is what Wayland needs.

`org.freedesktop.StatusNotifierItem` use screen coordinates, so I do not think we have a choice. Also, how it can be window-centric when there's no window.

**Comment by akien-mga** [comment]
So this got stale, can we reach a consensus?

It seems to me like we don't know yet whether FreeDesktop allows more than just screen coordinates for this API, so I feel like it might be best to merge this for now, and if we figure out another API would be better in the future, we can implement that and deprecate this one?

**Comment by bruvzg** [comment]
The primary purpose of this PR is positioning and displaying custom non-menu popups (which is not fully supported by DBus, it only provides click coordinates, not icon rect).

**Review by deralmas — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**